trl/trainer/grpo_trainer.py

```diff
@@ -300,6 +301,49 @@ def identity(x):
     return x
 
 
+def agg_loss(per_token_loss: torch.Tensor, loss_mask: torch.Tensor, loss_type: str, **kwargs) -> torch.Tensor:
+    if loss_type == "grpo":
+        loss = ((per_token_loss * loss_mask).sum(-1) / loss_mask.sum(-1).clamp(min=1.0)).mean()
+    elif loss_type == "bnpo":
+        loss = (per_token_loss * loss_mask).sum() / loss_mask.sum().clamp(min=1.0)
+    elif loss_type == "dr_grpo":
+        max_completion_length = kwargs.get("max_completion_length", 1)
+        loss = (per_token_loss * loss_mask).sum() / (per_token_loss.size(0) * max_completion_length)
+    else:
+        raise ValueError(f"Unknown loss type: {loss_type}")
+    return loss
+
+
+class AdaptiveEntropyController:
+    def __init__(self, min_ent_coef, max_ent_coef, delta_ent_coef, target_ent):
+        self.value = min_ent_coef
+        self.min_ent_coef = min_ent_coef
+        self.max_ent_coef = max_ent_coef
+        self.delta_ent_coef = delta_ent_coef
+        self.target_ent = target_ent
+
+    def __call__(self, ent):
+        """
+        Adjusts the entropy coefficient based on the current entropy.
+        Following https://github.com/SkyworkAI/Skywork-OR1/tree/main
+        Related post:
+        https://capricious-hydrogen-41c.notion.site/Skywork-Open-Reasoner-Series-1d0bc9ae823a80459b46c149e4f51680?pvs=25#1d1bc9ae823a801592a0c3891ea5328f
+
+        Args:
+            ent (`float`): Current entropy value.
+
+        Returns:
+            `float`: Adjusted entropy coefficient.
+        """
+        if ent < self.target_ent:
+            self.value += self.delta_ent_coef
+        else:
+            self.value -= self.delta_ent_coef
+
+        self.value = float(np.clip(self.value, self.min_ent_coef, self.max_ent_coef))
```

Comment:
let's avoid importing numpy just for this

```suggestion
        self.value = float(max(self.min_ent_coef, min(self.value, self.max_ent_coef)))
```